File: src/EditorFeatures/CSharp/Outlining/Outliners/DocumentationCommentOutliner.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Editor.Implementation.Outlining;
using Microsoft.CodeAnalysis.Shared.Extensions;
using Microsoft.CodeAnalysis.Text;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.Editor.CSharp.Outlining
{
    internal class DocumentationCommentOutliner : AbstractSyntaxNodeOutliner<DocumentationCommentTriviaSyntax>
    {
        private static string GetDocumentationCommentPrefix(DocumentationCommentTriviaSyntax documentationComment)
        {
            Contract.ThrowIfNull(documentationComment);

            var leadingTrivia = documentationComment.GetLeadingTrivia();
            var exteriorTrivia = leadingTrivia.Where(t => t.Kind() == SyntaxKind.DocumentationCommentExteriorTrivia)
                                              .FirstOrNullable();

            return exteriorTrivia != null ? exteriorTrivia.Value.ToString() : string.Empty;
        }

        private static string GetBannerText(DocumentationCommentTriviaSyntax documentationComment, CancellationToken cancellationToken)
        {
            // TODO: Consider unifying code to extract text from an Xml Documentation Comment (https://github.com/dotnet/roslyn/issues/2290)
            var summaryElement = documentationComment.ChildNodes().OfType<XmlElementSyntax>()
                                              .FirstOrDefault(e => e.StartTag.Name.ToString() == "summary");

            var prefix = GetDocumentationCommentPrefix(documentationComment);

            string text;
            if (summaryElement != null)
            {
                var sb = new StringBuilder(summaryElement.Span.Length);
                sb.Append(prefix);
                sb.Append(" <summary>");
                foreach (var node in summaryElement.ChildNodes())
                {
                    if (node is XmlTextSyntax)
                    {
                        var textTokens = ((XmlTextSyntax)node).TextTokens;
                        AppendTextTokens(sb, textTokens);
                    }
                    else if (node is XmlEmptyElementSyntax)
                    {
                        var e = (XmlEmptyElementSyntax)node;
                        foreach (var attribute in e.Attributes)
                        {
                            if (attribute is XmlCrefAttributeSyntax)
                            {
                                sb.Append(" ");
                                sb.Append(((XmlCrefAttributeSyntax)attribute).Cref.ToString());
                            }
                            else if (attribute is XmlNameAttributeSyntax)
                            {
                                sb.Append(" ");
                                sb.Append(((XmlNameAttributeSyntax)attribute).Identifier.Identifier.Text);
                            }
                            else if (attribute is XmlTextAttributeSyntax)
                            {
                                AppendTextTokens(sb, ((XmlTextAttributeSyntax)attribute).TextTokens);
                            }
                            else
                            {
                                Debug.Fail($"Unexpected XML syntax kind {attribute.Kind()}");
                            }
                        }
                    }
                }

                text = sb.ToString();
            }
            else
            {
                // If a summary element isn't found, use the first line of the XML doc comment.
                var syntaxTree = documentationComment.SyntaxTree;
                var spanStart = documentationComment.SpanStart;
                var line = syntaxTree.GetText(cancellationToken).Lines.GetLineFromPosition(spanStart);
                text = prefix + " " + line.ToString().Substring(spanStart - line.Start).Trim() + " " + CSharpOutliningHelpers.Ellipsis;
            }

            if (text.Length > CSharpOutliningHelpers.MaxXmlDocCommentBannerLength)
            {
                text = text.Substring(0, CSharpOutliningHelpers.MaxXmlDocCommentBannerLength) + " " + CSharpOutliningHelpers.Ellipsis;
            }

            return text;
        }

        private static void AppendTextTokens(StringBuilder sb, SyntaxTokenList textTokens)
        {
            foreach (var tk in textTokens)
            {
                var s = tk.ToString().Trim();
                if (s.Length > 0)
                {
                    sb.Append(" ");
                    sb.Append(s);
                }
            }
        }

        protected override void CollectOutliningSpans(
            DocumentationCommentTriviaSyntax documentationComment,
            List<OutliningSpan> spans,
            CancellationToken cancellationToken)
        {
            var startPos = documentationComment.FullSpan.Start;

            // The trailing newline is included in XmlDocCommentSyntax, so we need to strip it.
            var endPos = documentationComment.SpanStart + documentationComment.ToString().TrimEnd().Length;

            var span = TextSpan.FromBounds(startPos, endPos);

            spans.Add(new OutliningSpan(
                span,
                GetBannerText(documentationComment, cancellationToken),
                autoCollapse: true));
        }
    }
}

File: src/EditorFeatures/VisualBasic/Outlining/Outliners/DocumentationCommentOutliner.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports System.Text
Imports System.Threading
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Microsoft.CodeAnalysis.Editor.Implementation.Outlining

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.Outlining
    Friend Class DocumentationCommentOutliner
        Inherits AbstractSyntaxNodeOutliner(Of DocumentationCommentTriviaSyntax)

        Private Shared Function GetBannerText(documentationComment As DocumentationCommentTriviaSyntax, cancellationToken As CancellationToken) As String
            ' TODO: Consider unifying code to extract text from an Xml Documentation Comment (https://github.com/dotnet/roslyn/issues/2290)
            Dim summaryElement = documentationComment.Content _
                .OfType(Of XmlElementSyntax)() _
                .FirstOrDefault(Function(e) e.StartTag.Name.ToString = "summary")

            Dim text As String
            If summaryElement IsNot Nothing Then
                Dim sb As New StringBuilder(summaryElement.Span.Length)
                sb.Append("''' <summary>")
                For Each node In summaryElement.ChildNodes()
                    If node.Kind() = SyntaxKind.XmlText Then
                        Dim textNode = DirectCast(node, XmlTextSyntax)
                        Dim textTokens As SyntaxTokenList = textNode.TextTokens
                        AppendTextTokens(sb, textTokens)
                    ElseIf node.Kind() = SyntaxKind.XmlEmptyElement Then
                        Dim elementNode = DirectCast(node, XmlEmptyElementSyntax)
                        For Each attribute In elementNode.Attributes
                            If TypeOf attribute Is XmlCrefAttributeSyntax Then
                                sb.Append(" ")
                                sb.Append(DirectCast(attribute, XmlCrefAttributeSyntax).Reference.ToString())
                            ElseIf TypeOf attribute Is XmlNameAttributeSyntax Then
                                sb.Append(" ")
                                sb.Append(DirectCast(attribute, XmlNameAttributeSyntax).Reference.ToString())
                            ElseIf TypeOf attribute Is XmlAttributeSyntax Then
                                AppendTextTokens(sb, DirectCast(DirectCast(attribute, XmlAttributeSyntax).Value, XmlStringSyntax).TextTokens)
                            Else
                                Debug.Fail($"Unexpected XML syntax kind {attribute.Kind()}")
                            End If
                        Next
                    End If
                Next

                text = sb.ToString()
            Else
                ' If a summary element isn't found, use the first line of the XML doc comment.
                Dim span = documentationComment.Span
                Dim syntaxTree = documentationComment.SyntaxTree
                Dim line = syntaxTree.GetText(cancellationToken).Lines.GetLineFromPosition(span.Start)
                text = "''' " & line.ToString().Substring(span.Start - line.Start).Trim() & SpaceEllipsis
            End If

            If text.Length > MaxXmlDocCommentBannerLength Then
                text = text.Substring(0, MaxXmlDocCommentBannerLength) & SpaceEllipsis
            End If

            Return text
        End Function

        Private Shared Sub AppendTextTokens(sb As StringBuilder, textTokens As SyntaxTokenList)
            For Each token In textTokens.Where(Function(t) t.Kind = SyntaxKind.XmlTextLiteralToken)
                Dim s = token.ToString().Trim()
                If s.Length <> 0 Then
                    sb.Append(" ")
                    sb.Append(s)
                End If
            Next
        End Sub

        Protected Overrides Sub CollectOutliningSpans(documentationComment As DocumentationCommentTriviaSyntax, spans As List(Of OutliningSpan), cancellationToken As CancellationToken)
            Dim firstCommentToken = documentationComment.ChildNodesAndTokens().FirstOrNullable()
            Dim lastCommentToken = documentationComment.ChildNodesAndTokens().LastOrNullable()
            If firstCommentToken Is Nothing Then
                Return
            End If

            ' TODO: Need to redo this when DocumentationCommentTrivia.SpanStart points to the start of the exterior trivia.
            Dim startPos = firstCommentToken.Value.FullSpan.Start

            ' The trailing newline is included in DocumentationCommentTrivia, so we need to strip it.
            Dim endPos = lastCommentToken.Value.SpanStart + lastCommentToken.Value.ToString().TrimEnd().Length

            Dim fullSpan = TextSpan.FromBounds(startPos, endPos)

            spans.Add(
                CreateRegion(fullSpan, GetBannerText(documentationComment, cancellationToken), autoCollapse:=True))
        End Sub
    End Class
End Namespace
